odl/tomo/util/utility.py

```diff
@@ -172,6 +173,246 @@ def axis_rotation_matrix(axis, angle):
+        It is allowed to include a constant scaling but shouldn't have
+        strongly varying directional scaling (bad condition).
+
+    Returns
+    -------
+    transformed_vecs : tuple of `numpy.ndarray`, shape ``(ndim,)``
+        The transformed vectors. The first entry is (the transformed)
+        ``principal_vec``, followed by the transformed ``other_vecs``.
+        Thus the length of the tuple is ``len(other_vecs) + 1``.
+    matrix : `numpy.ndarray`, shape ``(ndim, ndim)``
+        Transformation matrix, either equal to ``matrix`` if provided,
+        or the computed rotation from ``principal_default`` to
+        (rescaled) ``principal_vec``.
+    """
+    transformed_vecs = []
+    principal_vec = np.asarray(principal_vec, dtype=float)
+    ndim = principal_vec.shape[0]
+
+    if matrix is None:
+        # Separate into dilation and rotation. The dilation is only used
+        # for comparison, not in the final matrix.
+        principal_default = np.asarray(principal_default, dtype=float)
+
+        pr_norm = np.linalg.norm(principal_vec)
+        pr_default_norm = np.linalg.norm(principal_default)
+
+        if pr_default_norm == 0.0 and pr_norm != 0.0:
+            raise ValueError('no transformation from {} to {}'
+                             ''.format(principal_default, principal_vec))
+        elif pr_norm == 0.0 and pr_default_norm != 0.0:
+            raise ValueError('transformation from {} to {} is singular'
+                             ''.format(principal_default, principal_vec))
+        elif pr_norm == 0.0 and pr_default_norm == 0.0:
+            dilation = 1.0
+        else:
+            dilation = (np.linalg.norm(principal_vec) /
+                        np.linalg.norm(principal_default))
+
+        # Determine the rotation part
+        if np.allclose(principal_vec, dilation * principal_default):
+            # Dilation only
+            matrix = np.eye(ndim)
+        else:
+            matrix = rotation_matrix_from_to(principal_default, principal_vec)
+
+        # This one goes straight in
+        transformed_vecs.append(principal_vec)
+
+    else:
+        matrix = np.asarray(matrix, dtype=float)
+        if matrix.shape != (ndim, ndim):
+            raise ValueError('matrix shape must be {}, got {}'
+                             ''.format((ndim, ndim), matrix.shape))
+
+        # Check matrix condition
+        svals = np.linalg.svd(matrix, compute_uv=False)
+        condition = np.inf if 0.0 in svals else svals[0] / svals[-1]
+        if condition > 1e6:
+            raise np.linalg.LinAlgError(
+                'matrix is badly conditioned: condition number is {}'
+                ''.format(condition))
+
+        transformed_vecs.append(matrix.dot(principal_vec))
+
+    for vec in other_vecs:
+        if vec is None:
+            transformed_vecs.append(None)
+        else:
+            transformed_vecs.append(matrix.dot(vec))
+
+    return tuple(transformed_vecs), matrix
```

Comment:
Turns out I don't use the returned matrix later so no need to return it in the first place.